crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,516 @@
+            bytes: ExternIO::from(payload),
+            signature: Signature::try_from(sig.to_vec())?,
+        })))
+        .await?;
+
+    match response {
+        AppResponse::ZomeCalled(response) => {
+            let response: serde_json::Value = response.decode()?;
+            serde_json::to_writer(std::io::stdout(), &response)?;
+            println!(); // Add newline
+        }
+        other => anyhow::bail!("Unexpected response while calling zome: {:?}", other),
+    }
+
+    Ok(())
+}
+
+fn generate_signing_credentials(
+    passphrase: BufRead,
+) -> anyhow::Result<(Auth, ed25519_dalek::SigningKey)> {
+    if unsafe { libsodium_sys::sodium_init() } < 0 {
+        anyhow::bail!("Failed to initialize libsodium");
+    }
+
+    let auth = load_or_create_auth()?;
+
+    let mut out = [0u8; 32];
+    if unsafe {
+        let read_guard = passphrase.read_lock();
+        libsodium_sys::crypto_pwhash(
+            out.as_mut_ptr(),
+            out.len() as u64,
+            std::ffi::CString::new(read_guard.as_ref())?.into_raw(),
+            read_guard.len() as u64,
+            auth.salt.as_ptr(),
+            libsodium_sys::crypto_pwhash_OPSLIMIT_INTERACTIVE as u64,
+            libsodium_sys::crypto_pwhash_MEMLIMIT_INTERACTIVE as usize,
+            libsodium_sys::crypto_pwhash_ALG_DEFAULT as i32,
+        ) != 0
+    } {
+        anyhow::bail!("Failed to derive key from password");
+    }
+
+    Ok((auth, ed25519_dalek::SigningKey::from_bytes(&out)))
+}
+
+async fn admin_port_from_connect_args(
+    connect_args: ConnectArgs,
+    admin_port: Option<u16>,
+) -> anyhow::Result<u16> {
+    // Use overridden admin port if provided, otherwise if the running argument was provided, use
+    // that, otherwise load the existing paths from the .hc file, filter by index and get the
+    // admin ports. If nothing is configured, load the `.hc` file from the current directory.
+    if let Some(admin_port) = admin_port {
+        Ok(admin_port)
+    } else if let Some(admin_port) = connect_args.running {
+        Ok(admin_port)
+    } else if let Some(index) = connect_args.index {
+        let paths = Existing {
+            existing_paths: vec![],
+            all: false,
+            last: false,
+            indices: vec![index as usize],
+        }
+        .load()?;
+
+        if let Some(admin_port) = get_admin_ports(paths).await?.first() {
+            Ok(*admin_port)
+        } else {
+            anyhow::bail!("No admin port found")
+        }
+    } else {
+        let paths = crate::save::load(std::env::current_dir()?)?;
+
+        if let Some(admin_port) = get_admin_ports(paths).await?.first() {
+            Ok(*admin_port)
+        } else {
+            anyhow::bail!("No admin port found")
+        }
+    }
+}
+
+struct AppClient {
+    ws_send: holochain_websocket::WebsocketSender,
+    _recv: WsPollRecv,
+}
+
+impl AppClient {
+    async fn try_new(admin_port: u16, installed_app_id: InstalledAppId) -> anyhow::Result<Self> {
```

Comment:
Code like this exists in several places. I'd like to extra it to common code at some point. You have to write code like this with the Rust client and it's in `hcterm` too.

Reply:
Circular dependency between repositories if we use it in the sandbox. But no, the client doesn't actually provide this code. It'd provide the "client" logic but it doesn't know how to allocate an app interface that matches the current client. I think that's common code but probably doesn't belong in a client. The client should just provide the ability to call Holochain and ideally shouldn't contain logic that is more in the "runtime" space.

So I think this belongs somewhere, maybe in a dedicated "runtime tools" crate. But not in the client and not here.